## ACTIVIDAD 3


*   Nombre: Gonzalo Gutierrez Daza
*   Curso: Mineria de datos



### 1.-Instalar bibliotecas

In [ ]:
import numpy as np  
import matplotlib.pyplot as plt  
import pandas as pd  
from itertools import combinations  

### 2.-Importar datos (base depurada) 🟠

Importamos desde Drive

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Convertimos a lista

In [517]:
data = np.load("/content/drive/MyDrive/Colab Notebooks/Mineria de datos/spotify.npy",allow_pickle=True)
data = data.tolist()

Mostramos cantidad de playlist

In [518]:
print(len(data))

10000


# 3. Definimos umbrales

In [519]:
min_support = 10 # Número mínimo de canciones a considerar en las reglas
confianza_min =0.015
sustento_min =1.5

**Modulo para calcular la frecuencia get_frecuent_itemset**




In [520]:
def get_frecuent_itemset(playlist,min_support):
    #calculando lista cocn elementos de minimo soporte
    all_list=[]
    for i in range(len(playlist)):
        if len(playlist[i])>min_support:
            all_list.append(playlist[i])

    #calculando la frecuencia de caniones unicas
    diccionario_frecuencias = {}
    for i in range(len(playlist)):
        for j in range(len(playlist[i])):
            palabra=playlist[i][j]
            if palabra in diccionario_frecuencias:
                diccionario_frecuencias[palabra] += 1
            else:
                diccionario_frecuencias[palabra] = 1

    print("-- FRECUENCIA HALLADA EXITOSAMENTE --")
    return all_list,diccionario_frecuencias


**Hallamos frecuencias y mostrar cantidad de canciones únicas**

In [521]:
list_playlists,Frec_songs=get_frecuent_itemset(data,min_support)
print("Existen ",len(Frec_cancion) ," canciones únicas")
print("Existen ",len(list_playlists) ," playlist con #elementos mayores al min_suport")

-- FRECUENCIA HALLADA EXITOSAMENTE --
Existen  132920  canciones únicas
Existen  9580  playlist con #elementos mayores al min_suport


**Visualizamos frecuencia de algunas canciones**

In [522]:
k=0
for palabra in Frec_cancion:
    frecuencia = Frec_cancion[palabra]
    print(f"La cancion -> '{palabra}' tiene una frecuencia de {frecuencia}")
    k=k+1
    if(k>10):break

La cancion -> 'Lose Control (feat. Ciara & Fat Man Scoop)' tiene una frecuencia de 73
La cancion -> 'Toxic' tiene una frecuencia de 128
La cancion -> 'Crazy In Love' tiene una frecuencia de 190
La cancion -> 'Rock Your Body' tiene una frecuencia de 87
La cancion -> 'It Wasn't Me' tiene una frecuencia de 221
La cancion -> 'Yeah!' tiene una frecuencia de 244
La cancion -> 'My Boo' tiene una frecuencia de 93
La cancion -> 'Buttons' tiene una frecuencia de 69
La cancion -> 'Say My Name' tiene una frecuencia de 268
La cancion -> 'Hey Ya! - Radio Mix / Club Mix' tiene una frecuencia de 212
La cancion -> 'Promiscuous' tiene una frecuencia de 166


# 3. Hallando soporte

Para el soporte tomaremos un umbral de 0.015 respecto a cada cancion esto para limitar y tener un menor data set, y ademas porque las canciones con un soporte bajo no seran relevantes para la creacion de reglar en la que nos basaremos

In [523]:
def hallar_soporte(dic_frec):
    diccionario_soporte = {}
    for palabra in dic_frec:
        #print(dic_frec[palabra])
        if(dic_frec[palabra]/len(data)>=confianza_min):
            diccionario_soporte[palabra]=dic_frec[palabra]/len(data)
    return diccionario_soporte

Hallamos el soporte de la primera iteracion

In [524]:
dict_soporte=hallar_soporte(Frec_cancion)
print(len(dict_soporte))

371


Generar nuevos item set en base a las combinaciones

In [525]:
def existe_arrX_en_arrY(X,Y):
    existe=True
    for i in range(len(X)):
        if X[i] in Y:
            existe =True
        else:
            existe=False
            break
    return existe
def get_frecuen_combinat(dict_soporte,playlist):
    songs_uni=[]
    list_tuplas_total=[]
    #tuplas_aux=[]
    #diccionario_frecuencias_tuplas_final = {}
    diccionario_frecuencias_tuplas_aux={}
    diccionario_soporte_tuplas={}
    combinacion_n=2
    for palabra in dict_soporte:
        songs_uni.append(palabra)
    while(True):
        #calculamos combinaciones de n
        print("INICIAMOS CALCULOS COMBINATORIOS")
        tuplas_aux=list(combinations(songs_uni, combinacion_n))
        #contamos freciencia de cada combiancion en las playlist
        for i in range(len(tuplas_aux)):
            for j in range(len(playlist)):
                if (existe_arrX_en_arrY(tuplas_aux[i],data[j])==True):
                    if (tuplas_aux[i] in diccionario_frecuencias_tuplas_aux) :
                        diccionario_frecuencias_tuplas_aux[tuple(tuplas_aux[i])] += 1
                    else:
                        diccionario_frecuencias_tuplas_aux[tuple(tuplas_aux[i])] = 1
        print(diccionario_frecuencias_tuplas_aux)  
        #print(diccionario_frecuencias_tuplas_aux)
        print("TERMINAMOS CONTEO DE FRECUENCIAS ")
        #aplicamos umbral
        for palabra in diccionario_frecuencias_tuplas_aux:
            if(diccionario_frecuencias_tuplas_aux[palabra]/len(data)>=0.015):
                diccionario_soporte_tuplas[palabra]=diccionario_frecuencias_tuplas_aux[palabra]/len(data)
            print(palabra,diccionario_frecuencias_tuplas_aux[palabra]/len(data))
        print(diccionario_soporte_tuplas)
        print("TERMINAMOS DE PLAICAR UMBRAL ")
        break
    return diccionario_soporte_tuplas

In [ ]:
a=["a","b","c","d"]
b=["c","a"]
print(existe_arrX_en_arrY(b,a))

True


In [ ]:
diccionario_frecuencias_tuplas_aux={}
song=["Crazy In Love",'Fake Love',"It Wasn't Me",'Antidote']
tuplas_aux=list(combinations(song, 2))
print(tuplas_aux)
for i in range(len(tuplas_aux)):
    for j in range(len(data)):
        if (existe_arrX_en_arrY(tuplas_aux[i],data[j])==True):
            if (tuplas_aux[i] in diccionario_frecuencias_tuplas_aux) :
                diccionario_frecuencias_tuplas_aux[tuple(tuplas_aux[i])] += 1
            else:
                diccionario_frecuencias_tuplas_aux[tuple(tuplas_aux[i])] = 1
print(diccionario_frecuencias_tuplas_aux)  

[('Crazy In Love', 'Fake Love'), ('Crazy In Love', "It Wasn't Me"), ('Crazy In Love', 'Antidote'), ('Fake Love', "It Wasn't Me"), ('Fake Love', 'Antidote'), ("It Wasn't Me", 'Antidote')]
{('Crazy In Love', 'Fake Love'): 13, ('Crazy In Love', "It Wasn't Me"): 28, ('Crazy In Love', 'Antidote'): 12, ('Fake Love', "It Wasn't Me"): 20, ('Fake Love', 'Antidote'): 45, ("It Wasn't Me", 'Antidote'): 17}


In [526]:
dict_soport_tuplas=get_frecuen_combinat(dict_soporte,data)

Se han truncado las últimas 5000 líneas del flujo de salida.
('Party In The U.S.A.', 'Drowning (feat. Kodak Black)') 0.0003
('Party In The U.S.A.', 'goosebumps') 0.0003
('Party In The U.S.A.', 'Slippery (feat. Gucci Mane)') 0.0001
('Party In The U.S.A.', 'm.A.A.d city') 0.0006
('Party In The U.S.A.', 'Juke Jam (feat. Justin Bieber & Towkio)') 0.0001
('Party In The U.S.A.', 'Bodak Yellow') 0.0005
('Party In The U.S.A.', 'Go Flex') 0.0001
('Party In The U.S.A.', 'Come Get Her') 0.0006
('Party In The U.S.A.', 'Deja Vu') 0.0003
('Party In The U.S.A.', 'No Type') 0.0007
('Party In The U.S.A.', 'Butterfly Effect') 0.0003
('Party In The U.S.A.', 'oui') 0.0008
('Party In The U.S.A.', 'Intro') 0.0003
('Party In The U.S.A.', 'Smile') 0.0008
('Party In The U.S.A.', 'You') 0.0003
('Party In The U.S.A.', 'Let Her Go') 0.0006
('Party In The U.S.A.', 'A Thousand Years') 0.0006
('Party In The U.S.A.', 'Down') 0.0035
('Party In The U.S.A.', 'Inside Out') 0.0002
('Party In The U.S.A.', 'Slide') 0.0005
(

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



Se han truncado las últimas 5000 líneas del flujo de salida.
('LOVE. FEAT. ZACARI.', 'Lose Yourself - Soundtrack Version') 0.0004
('LOVE. FEAT. ZACARI.', 'Gold Digger') 0.0012
('LOVE. FEAT. ZACARI.', 'Better Together') 0.0002
('LOVE. FEAT. ZACARI.', 'No Role Modelz') 0.0029
('LOVE. FEAT. ZACARI.', 'Me, Myself & I') 0.0008
('LOVE. FEAT. ZACARI.', 'See You Again (feat. Charlie Puth) - From "Furious 7"') 0.0006
('LOVE. FEAT. ZACARI.', 'Monster') 0.0003
('LOVE. FEAT. ZACARI.', 'Thunder') 0.0005
('LOVE. FEAT. ZACARI.', '7/11') 0.0003
('LOVE. FEAT. ZACARI.', 'Slow Motion') 0.0006
('LOVE. FEAT. ZACARI.', 'All I Want') 0.0003
('LOVE. FEAT. ZACARI.', 'Your Love') 0.0001
('LOVE. FEAT. ZACARI.', 'Alone') 0.0003
('LOVE. FEAT. ZACARI.', 'Africa') 0.0003
('LOVE. FEAT. ZACARI.', 'I Miss You') 0.0005
('LOVE. FEAT. ZACARI.', 'Stitches') 0.0002
('LOVE. FEAT. ZACARI.', 'Drops of Jupiter') 0.0002
('LOVE. FEAT. ZACARI.', 'My House') 0.0007
('LOVE. FEAT. ZACARI.', 'I Don’t Wanna Live Forever (Fifty Shades D

**Visualizacion de diccionario (Combinacion, Soporte)**

In [527]:
print(dict_soport_tuplas)

{('Closer', 'Let Me Love You'): 0.0159, ('goosebumps', 'XO TOUR Llif3'): 0.0152, ('goosebumps', 'HUMBLE.'): 0.0167, ('goosebumps', 'Congratulations'): 0.0155, ('XO TOUR Llif3', 'HUMBLE.'): 0.0204, ('XO TOUR Llif3', 'Mask Off'): 0.0163, ('XO TOUR Llif3', 'Congratulations'): 0.0179, ('iSpy (feat. Lil Yachty)', 'HUMBLE.'): 0.0155, ('iSpy (feat. Lil Yachty)', 'Bad and Boujee (feat. Lil Uzi Vert)'): 0.0152, ('iSpy (feat. Lil Yachty)', 'Congratulations'): 0.0162, ('DNA.', 'HUMBLE.'): 0.019, ('HUMBLE.', 'Mask Off'): 0.0204, ('HUMBLE.', 'Bad and Boujee (feat. Lil Uzi Vert)'): 0.0167, ('HUMBLE.', 'Bounce Back'): 0.0156, ('HUMBLE.', 'Congratulations'): 0.0214, ('Mask Off', 'Bad and Boujee (feat. Lil Uzi Vert)'): 0.0151, ('Mask Off', 'Congratulations'): 0.0162, ('Broccoli (feat. Lil Yachty)', 'Bad and Boujee (feat. Lil Uzi Vert)'): 0.0155, ('Broccoli (feat. Lil Yachty)', 'Caroline'): 0.0172, ('Broccoli (feat. Lil Yachty)', 'No Problem (feat. Lil Wayne & 2 Chainz)'): 0.0158, ('Bad and Boujee (feat

# 4. Creacion de reglas
## - Soporte
## - Confianza
## - Lift

In [537]:
def generate_association_rules(frequent_itemsets, confidence = 0.2, lift =1.5):
    tup=list(frequent_itemsets.keys())
    #print(len(tup))
    rules = [[0 for x in range(2)] for y in range(len(tup))] 
    for i in range(len(rules)):
        for j in range(2):
            #print(i,j)
            rules[i][j]=tup[i][j]
    # calculamos confianza
    dict_confianz=calcular_confianza(dict_soporte,dict_soport_tuplas)
    # calculamos lift
    dict_lift=calcular_lift(dict_soporte,dict_confianz)
    # validamos con confidence y lift
    tup=list(frequent_itemsets.keys())
    rules_final = []
    for k in range(len(tup)):
        #print(tup[k][0])
        #print(tup[k])
        conf_act=dict_confianz[tuple(tup[k])]
        lift_act=dict_lift[tuple(tup[k])]
        if (conf_act>=confidence) and  (lift_act>=lift):
            rules_final.append([tup[k],conf_act,lift_act])
    
    return rules,dict_confianz,dict_lift,rules_final
def calcular_confianza(dict_soporte,dict_soport_tuplas):
    dict_confianza={}
    tup=list(dict_soport_tuplas.keys())
    for k in range(len(tup)):
        #print(tup[k][0])
        #print(tup[k])
        dict_confianza[tup[k]]=dict_soport_tuplas[tuple(tup[k])]/(dict_soporte[tup[k][0]])
    return dict_confianza

def calcular_lift(dict_soporte,dict_confianz):
    dict_lift={}
    tup=list(dict_confianz.keys())
    for k in range(len(tup)):
        #print(tup[k][0])
        #print(tup[k])
        dict_lift[tup[k]]=dict_confianz[tuple(tup[k])]/(dict_soporte[tup[k][1]])
    return dict_lift

# Ejecución del algorítmo

In [538]:
rules,dict_confianz,dict_lift,dict_final=generate_association_rules(dict_soport_tuplas)

Visualizacion de dict_confianz y dict_lift

In [539]:
print(dict_confianz)
print(dict_lift)

{('Closer', 'Let Me Love You'): 0.20893561103810776, ('goosebumps', 'XO TOUR Llif3'): 0.4935064935064935, ('goosebumps', 'HUMBLE.'): 0.5422077922077921, ('goosebumps', 'Congratulations'): 0.5032467532467533, ('XO TOUR Llif3', 'HUMBLE.'): 0.6181818181818182, ('XO TOUR Llif3', 'Mask Off'): 0.49393939393939384, ('XO TOUR Llif3', 'Congratulations'): 0.5424242424242424, ('iSpy (feat. Lil Yachty)', 'HUMBLE.'): 0.45722713864306785, ('iSpy (feat. Lil Yachty)', 'Bad and Boujee (feat. Lil Uzi Vert)'): 0.44837758112094395, ('iSpy (feat. Lil Yachty)', 'Congratulations'): 0.47787610619469023, ('DNA.', 'HUMBLE.'): 0.8189655172413793, ('HUMBLE.', 'Mask Off'): 0.4340425531914894, ('HUMBLE.', 'Bad and Boujee (feat. Lil Uzi Vert)'): 0.3553191489361702, ('HUMBLE.', 'Bounce Back'): 0.33191489361702126, ('HUMBLE.', 'Congratulations'): 0.4553191489361702, ('Mask Off', 'Bad and Boujee (feat. Lil Uzi Vert)'): 0.4778481012658228, ('Mask Off', 'Congratulations'): 0.5126582278481012, ('Broccoli (feat. Lil Yachty

Visualizacion de reglas filtradas

In [540]:
print(dict_final)

[[('Closer', 'Let Me Love You'), 0.20893561103810776, 5.184506477372401], [('goosebumps', 'XO TOUR Llif3'), 0.4935064935064935, 14.954742227469499], [('goosebumps', 'HUMBLE.'), 0.5422077922077921, 11.53633600442111], [('goosebumps', 'Congratulations'), 0.5032467532467533, 12.27431105479886], [('XO TOUR Llif3', 'HUMBLE.'), 0.6181818181818182, 13.152804642166345], [('XO TOUR Llif3', 'Mask Off'), 0.49393939393939384, 15.630993479094741], [('XO TOUR Llif3', 'Congratulations'), 0.5424242424242424, 13.229859571322985], [('iSpy (feat. Lil Yachty)', 'HUMBLE.'), 0.45722713864306785, 9.7282369924057], [('iSpy (feat. Lil Yachty)', 'Bad and Boujee (feat. Lil Uzi Vert)'), 0.44837758112094395, 12.701914479346854], [('iSpy (feat. Lil Yachty)', 'Congratulations'), 0.47787610619469023, 11.655514785236347], [('DNA.', 'HUMBLE.'), 0.8189655172413793, 17.424798239178283], [('HUMBLE.', 'Mask Off'), 0.4340425531914894, 13.735523835173714], [('HUMBLE.', 'Bad and Boujee (feat. Lil Uzi Vert)'), 0.35531914893617

# RESULTADOS

## El modo de visualizar estos datos se visualizaran y ordenaran en base a la lift ya que le daremos importancia y sus antecedentes, de esta forma las pondremos de orden decreciente y obtendremos las mejorres puntuadas.

##Visualizacion mendiante data frame

In [542]:
association_rulesdf = pd.DataFrame(dict_final, columns=['X','Confianza','Lift'])
association_rulesdf

X  Confianza       Lift
0                           (Closer, Let Me Love You)   0.208936   5.184506
1                         (goosebumps, XO TOUR Llif3)   0.493506  14.954742
2                               (goosebumps, HUMBLE.)   0.542208  11.536336
3                       (goosebumps, Congratulations)   0.503247  12.274311
4                            (XO TOUR Llif3, HUMBLE.)   0.618182  13.152805
5                           (XO TOUR Llif3, Mask Off)   0.493939  15.630993
6                    (XO TOUR Llif3, Congratulations)   0.542424  13.229860
7                  (iSpy (feat. Lil Yachty), HUMBLE.)   0.457227   9.728237
8   (iSpy (feat. Lil Yachty), Bad and Boujee (feat...   0.448378  12.701914
9          (iSpy (feat. Lil Yachty), Congratulations)   0.477876  11.655515
10                                    (DNA., HUMBLE.)   0.818966  17.424798
11                                (HUMBLE., Mask Off)   0.434043  13.735524
12     (HUMBLE., Bad and Boujee (feat. Lil Uzi Vert))   0.355319  10.065698
13                             (HUMBLE., Bounce Back)   0.331915  11.063830
14                         (HUMBLE., Congratulations)   0.455319  11.105345
15    (Mask Off, Bad and Boujee (feat. Lil Uzi Vert))   0.477848  13.536773
16                        (Mask Off, Congratulations)   0.512658  12.503859
17  (Broccoli (feat. Lil Yachty), Bad and Boujee (...   0.385572  10.922723
18            (Broccoli (feat. Lil Yachty), Caroline)   0.427861  10.831916
19  (Broccoli (feat. Lil Yachty), No Problem (feat...   0.393035  12.168261
20  (Bad and Boujee (feat. Lil Uzi Vert), Bounce B...   0.478754  15.958451
21    (Bad and Boujee (feat. Lil Uzi Vert), Caroline)   0.433428  10.972855

##Por confianza

In [548]:
association_rulesdf_confianza=association_rulesdf.sort_values('Confianza',ascending=False)
association_rulesdf_confianza.head(10)

X  Confianza       Lift
10                                    (DNA., HUMBLE.)   0.818966  17.424798
4                            (XO TOUR Llif3, HUMBLE.)   0.618182  13.152805
6                    (XO TOUR Llif3, Congratulations)   0.542424  13.229860
2                               (goosebumps, HUMBLE.)   0.542208  11.536336
16                        (Mask Off, Congratulations)   0.512658  12.503859
3                       (goosebumps, Congratulations)   0.503247  12.274311
5                           (XO TOUR Llif3, Mask Off)   0.493939  15.630993
1                         (goosebumps, XO TOUR Llif3)   0.493506  14.954742
20  (Bad and Boujee (feat. Lil Uzi Vert), Bounce B...   0.478754  15.958451
9          (iSpy (feat. Lil Yachty), Congratulations)   0.477876  11.655515

##Por Lift

In [547]:
association_rulesdf_lift=association_rulesdf.sort_values('Lift',ascending=False)
association_rulesdf_lift.head(10)

X  Confianza       Lift
10                                    (DNA., HUMBLE.)   0.818966  17.424798
20  (Bad and Boujee (feat. Lil Uzi Vert), Bounce B...   0.478754  15.958451
5                           (XO TOUR Llif3, Mask Off)   0.493939  15.630993
1                         (goosebumps, XO TOUR Llif3)   0.493506  14.954742
11                                (HUMBLE., Mask Off)   0.434043  13.735524
15    (Mask Off, Bad and Boujee (feat. Lil Uzi Vert))   0.477848  13.536773
6                    (XO TOUR Llif3, Congratulations)   0.542424  13.229860
4                            (XO TOUR Llif3, HUMBLE.)   0.618182  13.152805
8   (iSpy (feat. Lil Yachty), Bad and Boujee (feat...   0.448378  12.701914
16                        (Mask Off, Congratulations)   0.512658  12.503859

# Explicacion reglas obtenidas

## Regla 1
            X 	           Confianza 	Lift
       (DNA., HUMBLE.) 	   0.818966 	17.424798

    Ambas canciones son escuchadas en ritmos como hip hop y poseen confianza y lift muy altos por lo cual hay una gran posibilidad de esucharlos juntos


## Regla 2
            X 	                         Confianza 	Lift
       (goosebumps, Congratulations) 	0.503247 	12.274311

    Ambas canciones son escuchadas en ritmos como rap y poseen un lift por lo que poseen una aparicion constante entre ambas.

## Regla 3
           
             X 	                                            Confianza 	Lift
    (Mask Off, Bad and Boujee (feat. Lil Uzi Vert)) 	0.477848 	13.536773

    Aqui tenemos 2 canciones con un lift alto, ademas que en zonas de ciertos lugares se suelete escuchar este genero urbano

## Regla 4
            X 	                                        Confianza 	Lift
    (Bad and Boujee (feat. Lil Uzi Vert), Caroline) 	0.433428 	10.972855

    Ambas canciones son de tirmo urbano aunque a tirmos disitnos esto puede ser un indicador de que no esta entre las mejores reglas que se encontro